### PR TITLE
Add Pow, Ln, and Exp functions

### PR DESCRIPTION
**Pull request by sam-vdp**
Add Exp and Ln functions and Pow based on the combination of both. Ln uses Newton's method to approximate the solution, this code was ported from libfixmath, which your license already covers. The Exp function uses a power series

### CODE REVIEW

File: src/Fix64.cs [hunk 1]
```diff
@@ -359,6 +368,94 @@ public partial struct Fix64 : IEquatable<Fix64>, IComparable<Fix64> {
             return x.m_rawValue <= y.m_rawValue;
         }
 
+        public static Fix64 Exp(Fix64 x)
+        {
+            if (x.RawValue == 0) return One;
+            if (x == One) return E;
+            if (x >= LnMax) return MaxValue;
```
Comment (asik) [inline src/Fix64.cs]: What are the checks against LnMax and LnMin for? The tests still pass without. Optimization?

**Review by asik — CHANGES_REQUESTED** [review]
Thank you so much for this contribution! I have made some comments here.

File: tests/Fix64Tests.cs [hunk 2]
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
...
+        public void Ln()
+        {
+            for (int j = 0; j < m_testCases.Length; ++j)
+            {
+                var b = Fix64.FromRaw(m_testCases[j]);
+
+                if (b <= Fix64.Zero)
+                    continue;
+
+                if (b > Fix64.MaxValue / (Fix64)100)
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply tests/Fix64Tests.cs]: Probably a leftover from a previous algorithm I tried that didn't handle large arguments well. Removed.

File: tests/Fix64Tests.cs [hunk 3]
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
...
+                    continue;
+
+                // Reduced precision requirements for very small values
+                double maxDelta = b < (Fix64)0.000001m ? 0.1 : (double)(Fix64.Precision * 10);
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply tests/Fix64Tests.cs]: The worst delta actually seen from the test data is 0.011, I've adjusted the test to reflect that. My guess would be that the errors of the argument scaling (repeated multiplication with E^4) add up and taint the result?

File: tests/Fix64Tests.cs [hunk 4]
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
...
+        }
+
+        [Fact]
+        public void Pow()
+        {
+            for (int i = 0; i < m_testCases.Length; ++i)
+            {
+                var b = Fix64.FromRaw(m_testCases[i]);
+
+                if (b < Fix64.Zero)
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply tests/Fix64Tests.cs]: Done.

File: src/Fix64.cs (hunk 1 above)
Comment (sam-vdp) [reply src/Fix64.cs]: Yes. I figured if we know precisely where the argument range of the function ends, might as well test for it, but I guess if the early bailout test rarely succeeds it's a waste of cycles. I'm fine with removing it, your call.

File: tests/Fix64Tests.cs [hunk 5]
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
...
+                    continue;
+                if (b > Fix64.MaxValue / (Fix64)100)
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply tests/Fix64Tests.cs]: For very large and very small bases the precision becomes bad or terrible, mostly because Ln isn't very precise anymore.
Example:
`Pow(0.0000000002328306436538696289, -0.5) = expected 65536 but got 65901.3926025061`
Or somewhat pathological cases like this where it really falls apart:
`Pow(2147483647.9999999997671693563, 0.9999999997671693563461303711) = expected 2147483637.25622 but got 1426713511.50211`

Finally, maybe the error should be measured relatively, instead of absolutely? 
`Pow(2147483647.9999999997671693563, 0.7911919844336807727813720703) = expected 24173977.2437025 but got 24173942.9738885`

I don't see a way to fully fix these cases, but also describing the exact limitations is somewhat difficult. I have refactored the test somewhat and am looking forward to your comments.

File: tests/Fix64Tests.cs [hunk 6]
```diff
@@ -361,6 +361,78 @@ public void Sqrt()
...
         }
 
+        [Fact]
+        public void Ln()
+        {
+            for (int j = 0; j < m_testCases.Length; ++j)
+            {
+                var b = Fix64.FromRaw(m_testCases[j]);
+
+                if (b <= Fix64.Zero)
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply tests/Fix64Tests.cs]: Done.

File: src/Fix64.cs (hunk 1 above)
Comment (asik) [reply src/Fix64.cs]: It's a good optimization I think, just wanted to make sure.

File: tests/Fix64Tests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (asik) [reply tests/Fix64Tests.cs]: So, the test skips half the range [0..MaxValue], and in the range ]1.. MaxValue/2] the delta is artificially reduced to a fraction of itself. I'm not sure I understand the rationale behind this? In any case I'm not comfortable with an operation that's effectively untested over a large range of inputs and boundary conditions (even though I'm guilty of that myself e.g. Tan - I do document that it's not tested well though).

If Ln is the culprit then let's find a more accurate Ln. You got me reading on the topic :) I think an accurate base-2 logarithm for a 64-bit integer would be a good start. https://graphics.stanford.edu/~seander/bithacks.html#IntegerLog has ideas, https://stackoverflow.com/questions/11376288/fast-computing-of-log2-for-64-bit-integers as well. 

Once we have that, Fix64 is just a long divided by 2^32, and log(x/y) is log(x) - log(y), so log2 of a Fix64 is log2(RawValue) - 32 (right?)

Once we have an accurate log2, ln x is just log2 x/log2 e.

**Review by asik — CHANGES_REQUESTED** [review]

File: src/Fix64.cs [hunk 7]
```diff
@@ -359,6 +368,94 @@ public partial struct Fix64 : IEquatable<Fix64>, IComparable<Fix64> {
...
+
+            // The power-series converges much faster on positive values
+            // and exp(-x) = 1/exp(x).
+            bool neg = (x.RawValue < 0);
+            if (neg) x = -x;
+
+            Fix64 result = x + One;
+            Fix64 term = x;
+
+            for (int i = 2; i < 40; i++)
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply src/Fix64.cs]: WRT to your comment below: Yes, a better Ln implementation would be great! However, in your comment below, I think it's half of the solution. The code in the links calculates only the integer part of the log2, but we'll still need the fractional part, right?
Wikipedia has an algorithm for that, but I would assume comes with a similar speed and precision issues:
https://en.wikipedia.org/wiki/Binary_logarithm#Iterative_approximation

This one looks promising, I'll have a look at that:
https://github.com/dmoulding/log2fix

**Comment by sam-vdp** [comment]
Thank you very much for pushing back on the crappy previous solution and for your very good idea of using log2. I've added Log2 and Ln based on it, which don't have the previous precision issues.

Furthermore I've removed Exp and added Pow2, which splits the exponentiation into an integer and fractional part and handles large values much better, while maintaining precision for smaller values.

Please note that the Pow test still has some heuristics regarding the expected precision. Firstly, for large results the absolute errors become larger, which needs to be compensated. Secondly, some test cases are slightly pathological. E.g. `Pow(1-Fix64.FromRaw(1), Fix64.MaxValue)` is evaluated by doing `Pow2(Fix64.MaxValue * Fix64.Log2(1-Fix64.FromRaw(1))`, which becomes `Pow2(Fix64.MaxValue * Fix64.Precision)` which of course is not very precise.

File: tests/Fix64Tests.cs [hunk 8]
```diff
@@ -428,12 +445,12 @@ public void Pow()
                     {
                         var expected = e == Fix64.Zero ? 1 : b == Fix64.Zero ? 0 : Math.Min(Math.Pow((double)b, (double)e), (double)Fix64.MaxValue);
```
Comment (asik) [inline tests/Fix64Tests.cs]: This is hiding at least one bug: it ignores the case Pow(0, negative power) which should give MaxValue (the closest we have to infinity).

File: tests/Fix64Tests.cs [hunk 9]
```diff
@@ -428,12 +445,12 @@ public void Pow()
...
 
-                        if (expected > (double)Fix64.MaxValue / 2)
-                            continue;
+                        // Absolute precision deteriorates with large result values, take this into account
+                        // Similarly, large exponents reduce precision, even if result is small.
+                        double maxDelta = Math.Abs((double)e) > 100000000 ? 0.5 : expected > 100000000 ? 10 : expected > 1000 ? 0.5 : 0.00001;
```
Comment (asik) [inline tests/Fix64Tests.cs]: Actually if the "real" Pow(x, y) > MaxValue then you expect exactly MaxValue with 0 delta.

File: src/Fix64.cs [hunk 10]
```diff
@@ -382,66 +395,68 @@ public static Fix64 Exp(Fix64 x)
...
             }
 
+            result = FromRaw(result.m_rawValue << integerPart);
             if (neg) result = One / result;
 
             return result;
         }
 
-        public static Fix64 Ln(Fix64 x)
+        public static Fix64 Log2(Fix64 x)
```
Comment (asik) [inline src/Fix64.cs]: I wouldn't expose Log2 as public API. A more useful API (which can be a separate PR) would be Log(a, newBase) like System.Math offers.

File: src/Fix64.cs [hunk 11]
```diff
@@ -382,66 +395,68 @@ public static Fix64 Exp(Fix64 x)
...
+                    y += b;
+                }
+                b >>= 1;
+            }
 
-            return guess + (Fix64)scaling;
+            return Fix64.FromRaw(y);
+        }
 
+        public static Fix64 Ln(Fix64 x)
```
Comment (asik) [inline src/Fix64.cs]: All public APIs need a ///summary. I try to stay close to System.Math in descriptions. Also discussion of accuracy if it's not accurate down to Fix64.Precision (see for ex. Sin in the latest master - I just updated it yesterday).

**Review by asik — CHANGES_REQUESTED** [review]

File: tests/Fix64Tests.cs (hunk 8 above)
Comment (sam-vdp) [reply tests/Fix64Tests.cs]: Mathematically, 0^x (x<0) is undefined, so I would argue 0, MaxValue or Infinity are pretty much equally wrong. Windows Calculator for example returns 0. I've added an exception for this case, after all, the division operator also throws instead of return MaxValue.

File: tests/Fix64Tests.cs (hunk 9 above)
Comment (sam-vdp) [reply tests/Fix64Tests.cs]: That would be nice, but actually the Pow approximation is still subject to the limitations that increase the maxDelta to 0.5 before. I.e. even if the expected value is MaxValue, the actual result may be slightly below, but inside the 0.5 delta.

File: tests/Fix64Tests.cs [hunk 12]
```diff
@@ -387,18 +410,19 @@ public void Ln()
         }
 
         [Fact]
-        public void Exp()
+        public void Pow2()
         {
-            double maxDelta = (double)(Fix64.Precision * 10);
+            var test = (double)Fix64.Pow2((Fix64)3);
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply tests/Fix64Tests.cs]: Remove

File: src/Fix64.cs [hunk 13]
```diff
@@ -99,6 +98,14 @@ public partial struct Fix64 : IEquatable<Fix64>, IComparable<Fix64> {
             return hasFractionalPart ? Floor(value) + One : value;
         }
 
+        /// <summary>
+        /// Returns the fractional part of the specified number.
+        /// </summary>
+        public static Fix64 FractionalPart(Fix64 value)
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply src/Fix64.cs]: Removed

File: src/Fix64.cs [hunk 14]
```diff
@@ -368,12 +375,18 @@ public partial struct Fix64 : IEquatable<Fix64>, IComparable<Fix64> {
             return x.m_rawValue <= y.m_rawValue;
         }
 
-        public static Fix64 Exp(Fix64 x)
+        public static Fix64 Pow2(Fix64 x)
```
[the remark this answers is not in the document]
Comment (sam-vdp) [reply src/Fix64.cs]: The thing with Pow2 and Log2 is, that I would still like to have dedicated tests for them, not just their combined results. With the tests in the separate assembly, I don't see how that would be possible without making them public?

[the remark this answers is not in the document]
Comment (asik) [reply src/Fix64.cs]: Good point. Make it internal and add a InternalsVisibleTo attribute to the assembly in an AssemblyInfo.cs file.

[the remark this answers is not in the document]
Comment (sam-vdp) [reply src/Fix64.cs]: That's a neat trick. Done.

**Comment by asik** [comment]
I've made a few stylistic changes locally and merged. Thank you for this contribution!
